src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -100,105 +69,117 @@ protected override Expression VisitBinary(BinaryExpression expression)
         /// An '= ANY' expression or null.
         /// </returns>
         [CanBeNull]
-        protected virtual Expression VisitEqualsAny([NotNull] SubQueryExpression expression)
-        {
-            var subQueryModel = expression.QueryModel;
-            var fromExpression = subQueryModel.MainFromClause.FromExpression;
+        protected override Expression VisitSubQuery(SubQueryExpression expression)
+            => base.VisitSubQuery(expression) ?? VisitArraySubQuery(expression);
 
-            var properties = MemberAccessBindingExpressionVisitor.GetPropertyPath(
-                fromExpression, _queryModelVisitor.QueryCompilationContext, out _);
+        /// <inheritdoc />
+        protected override Expression VisitUnary(UnaryExpression expression)
+        {
+            // Character literals are represented as integer literals by the expression tree.
+            // So `myString[0] == 'T'` looks like `((int)myString[0]) == 84`.
+            // Since `substr` returns `text`, not `char` or `int`, wrap it in `ascii`.
+            if (expression.NodeType == ExpressionType.Convert &&
+                Visit(expression.Operand) is IndexExpression index &&
+                index.Object.Type == typeof(string))
+                return new SqlFunctionExpression("ascii", typeof(int), new[] { index });
+
+            return base.VisitUnary(expression);
+        }
 
-            if (properties.Count == 0)
-                return null;
-            var lastPropertyType = properties[properties.Count - 1].ClrType;
-            if (lastPropertyType.IsArray && lastPropertyType.GetArrayRank() == 1 && subQueryModel.ResultOperators.Count > 0)
+        /// <inheritdoc />
+        /// <remarks>
+        /// https://github.com/aspnet/EntityFrameworkCore/blob/release/2.2/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L1077-L1144
+        /// </remarks>
+        protected override Expression VisitExtension(Expression expression)
+        {
+            switch (expression)
+            {
+            case ArrayAnyAllExpression e:
```

Comment:
I'm confused... `NpgsqlArrayFragmentTranslator` now constructs and returns `ArrayAnyAllExpression`s, are they expected to find their way back here in the SqlTranslatingExpressionVisitor? Is this because we assume the fragment translators execute before this method? Either way, the whole point of having the FragmentTranslator is to cleanly encapsulate all the array logic in one place, but this seems to spread it back around...

The situation seems to be similar with the other custom expression we have, but I may be misunderstanding completely.

Reply:
I based this on the [implementation in SqlTranslatingExpressionVisitor](https://github.com/aspnet/EntityFrameworkCore/blob/8e7402701ea1416b7fe55e1863ae65b3147c02c7/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L1091-L1158) which does similar work on `NodeType.Extension` expressions defined upstream.

The fragment translators [always run first](https://github.com/aspnet/EntityFrameworkCore/blob/8e7402701ea1416b7fe55e1863ae65b3147c02c7/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L103-L105), so the idea is to package up a bunch of complicated nodes into one custom expression that we can easily identify and handle later on in the normal translator workflow.

It does spread around _some_ logic, but this is still advantageous for the `NpgsqlSqlTranslatingExpressionVisitor`, because we're just performing normal expression visitor operations on `ArrayAnyAllExpression` (e.g. visit members, return original or modified).

_Note we do have one remaining piece of complicated `ArrayAnyAllExpression` logic outside the fragment translator, which involves translating a `ContainsResultOperator`. I want to refactor that into a fragment translator too, but I'd like to deal with that in a later PR._